source/common/upstream/conn_pool_map_impl.h

```diff
@@ -6,30 +6,42 @@ namespace Envoy {
 namespace Upstream {
 
 template <typename KEY_TYPE, typename POOL_TYPE>
-ConnPoolMap<KEY_TYPE, POOL_TYPE>::ConnPoolMap(Envoy::Event::Dispatcher& dispatcher)
-    : thread_local_dispatcher_(dispatcher) {}
+ConnPoolMap<KEY_TYPE, POOL_TYPE>::ConnPoolMap(Envoy::Event::Dispatcher& dispatcher,
+                                              absl::optional<uint64_t> max_size)
+    : thread_local_dispatcher_(dispatcher), max_size_(max_size) {}
 
 template <typename KEY_TYPE, typename POOL_TYPE>
 ConnPoolMap<KEY_TYPE, POOL_TYPE>::~ConnPoolMap() = default;
 
 template <typename KEY_TYPE, typename POOL_TYPE>
-POOL_TYPE& ConnPoolMap<KEY_TYPE, POOL_TYPE>::getPool(KEY_TYPE key, const PoolFactory& factory) {
+typename ConnPoolMap<KEY_TYPE, POOL_TYPE>::OptPoolRef
+ConnPoolMap<KEY_TYPE, POOL_TYPE>::getPool(KEY_TYPE key, const PoolFactory& factory) {
   Common::AutoDebugRecursionChecker assert_not_in(recursion_checker_);
   // TODO(klarose): Consider how we will change the connection pool's configuration in the future.
   // The plan is to change the downstream socket options... We may want to take those as a parameter
   // here. Maybe we'll pass them to the factory function?
-  auto inserted = active_pools_.emplace(key, nullptr);
-
-  // If we inserted a new element, create a pool and assign it to the iterator. Tell it about any
-  // cached callbacks.
-  if (inserted.second) {
-    inserted.first->second = factory();
-    for (const auto& cb : cached_callbacks_) {
-      inserted.first->second->addDrainedCallback(cb);
+  auto pool_iter = active_pools_.find(key);
+  if (pool_iter != active_pools_.end()) {
+    return std::ref(*(pool_iter->second));
+  }
+
+  // We need a new element. If we can, create a pool and add it to the map. Tell it about any cached
+  // callbacks.
+  if (max_size_.has_value() && size() >= max_size_) {
```

Comment:
nit: `max_size_. value()` (I can never remember if without it means the value or just whether it has a value)